File: Leetcode/Matrix Block Sum/Matrix Block Sum.java
class Solution {
    public int[][] matrixBlockSum(int[][] mat, int K) {
        //Finding the sizes of rows and columns
        int m = mat.length;
        int n = mat[0].length;
        //This matrix will contain our answer
        int[][] ans = new int[m][n];
        //This matrix will contain prefix sums
        int[][] prefix = new int[m][n];
        //Calculating prefix sums for every i and j
        for(int i = 0; i < m; i++) {
            int sum = 0;
            for(int j = 0; j < n; j++) {
                sum += mat[i][j];
                prefix[i][j] = sum;
            }
        }
        //Now, we will calculate valid range of rows and columns first
        //Then we add prefix sums to answer for row range and column's upper bound
        //And if column's lower bound is not 0, we substract those prefix sums
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                //Finding upper and lower bound for rows
                int r1 = Math.max(0, i - K);
                int r2 = Math.min(m - 1, i + K);
                //Finding upper and lower bounds for columns
                int c1 = Math.max(0, j - K); 
                int c2 = Math.min(n - 1, j + K);
                //Finding answer matrix
                for(int p = r1; p <= r2; p++) {
                    ans[i][j] += prefix[p][c2];
                    if(c1 != 0) {
                        ans[i][j] -= prefix[p][c1 - 1];
                    }
                }
            }
        }
        return ans;
    }
}

File: Leetcode/Matrix Block Sum/Matrix_block_sum.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
		vector<vector<int>> matrixBlockSum(vector<vector<int>>& mat, int K) 
        {
            //declaration of matrix
            vector<vector<int>> sum(mat.size()+1,vector<int>(mat[0].size()+1,0));
            
            //loop for calculating sum of all elements accordingly
            for(int i=1;i<sum.size();i++)
            {
                for(int j=1;j<sum[i].size();j++)
                {
                        sum[i][j]=sum[i-1][j]-sum[i-1][j-1]+sum[i][j-1]+mat[i-1][j-1];
                }
            }
            
            //variables for size of input matrix 
            int m=mat.size();
            int n=mat[0].size();
            
            //loop for storing sum of bocks in their right position according to the question
            for(int i=0;i<m;i++)
            {
                for(int j=0;j<n;j++) 
                {
                    int r1= max(0,i-K);
                    int c1= max(0,j-K); 
                    int r2= min(i+K,m-1); 
                    int c2= min(j+K,n-1); 
                    mat[i][j]=sum[r2+1][c2+1]-sum[r1][c2+1]-sum[r2+1][c1]+sum[r1][c1];
                }
            }
            
            //returning matrix 
            return mat;
        }
};